src/grimp/adaptors/_layers.py

```diff
@@ -52,11 +53,24 @@ class _RustPackageDependency(TypedDict):
     routes: tuple[_RustRoute, ...]
 
 
-def _layers_to_levels(layers: Sequence[Union[str, set[str]]]) -> tuple[set[str], ...]:
+def _layers_to_levels(layers: Sequence[Union[str, set[str], Level]]) -> tuple[Level, ...]:
```

Comment:
This change to the API feels quite significant, and is the only thing that concerns me about this PR. As you've probably noticed, up until now the API just uses native Python datatypes and I think that has worked well generally, though it is being stretched a little with the more complex return types of `find_illegal_dependencies_for_layers` and `get_import_details`. I do really like the way we haven't needed any custom datatypes so far, but I'm starting to wonder if it's time to move beyond this as the API grows in complexity.

For me, passing it in as a dict is `dict(independent=False, layers={"green", "blue"})` probably crosses a boundary: a custom datatype would be simpler here. That said, I do really like the approach of backward compatibility you've taken. I've been trying to think of alternatives.

One option (let's call it Option A) would be to define `Level` as a custom class (not just a `TypedDict`). The API might look like this:

```
dependencies = graph.find_illegal_dependencies_for_layers(
    layers=(
        "red",
        grimp.Level({"green", "blue"}, independent=False),
        "yellow",
    ),
)
```

Notice `independent` is moved to the end and the layers have been made a positional argument, for terseness.

Another even terser option (Option B) would be to make all the layers positional so we don't need to pass them in as a set, i.e. `grimp.Level("green", "blue", independent=False)`.

With either of these options, I think it would be nice to make `Level` the canonical datatype passed to the function, with the `str` and `set[str]` options being conveniences that are converted to `Level` objects straight away. So we could also call the function like this:

```
dependencies = graph.find_illegal_dependencies_for_layers(
    layers=(
        grimp.Level("red"),
        grimp.Level("green", "blue", independent=False),
        grimp.Level("yellow", "orange"),
    ),
)
```

If we did that then it would make sense to annotate `layers` as `Level | str | set[str]`, to make level more prominent. Or possibly we should even rename it to `levels` (we'd need to figure out backward compatibility though).

Another option (Option C) I considered was to use the decorator pattern just to wrap the set:

```
dependencies = graph.find_illegal_dependencies_for_layers(
    layers=(
        "red",
        grimp.dependent({"green", "blue"}),
        "yellow",
    ),
)
```

This has the advantage of being very terse, though I do wonder if it is potentially confusing.

I'm leaning towards Option B, interested to hear your thoughts.

Reply:
I was originally thinking to introduce a class for this, a bit like you're suggesting. I had decided I wouldn't mainly because of the internal rust function - when calling the rust function it seemed easiest to work with a primative dict rather than a more complex type. Maybe there is a way to pass complex types to rust via `pyo3` though (do you happen to know?). Anyways, this internal decision shouldn't affect the public API of the python package, I agree.

I think option B works for me. 

I wonder about the naming - is `grimp.Level` meaningful enough? Or should it be something more verbose but explicit e.g. `grimp.LayeredArchitectureLevel`? What do you think?

Reply:
Should this `Level` type (or whatever we decide to name it) be a `ValueObject` https://github.com/seddonym/grimp/blob/master/src/grimp/domain/valueobjects.py?

Reply:
> when calling the rust function it seemed easiest to work with a primative dict rather than a more complex type

Agreed. It's fine to convert the object to a dict in Python before passing to Rust, that's just an internal implementation detail that can be easily changed.

Reply:
Updated in the option B fashion 🤞